modules/iio-als.c

```diff
@@ -0,0 +1,195 @@
+					error->message);
+				g_clear_pointer(&ret, g_variant_unref);
+				return false;
+			}
+			g_clear_pointer(&ret, g_variant_unref);
+
+			int ilux = iio_als_get_light_value(iio_proxy);
+			(void)execute_datapipe(&light_sensor_pipe, GINT_TO_POINTER(ilux), USE_INDATA, CACHE_INDATA);
+		} else if (!claim && claimed) {
+			mce_log(LL_DEBUG, "%s: ReleaseLight", MODULE_NAME);
+			ret =
+			    g_dbus_proxy_call_sync(iio_proxy, "ReleaseLight", NULL, G_DBUS_CALL_FLAGS_NONE, -1, NULL,
+						   &error);
+			if (!ret && !g_error_matches(error, G_IO_ERROR, G_IO_ERROR_CANCELLED)) {
+				mce_log(LL_WARN, "%s: failed to relese ambient light sensor %s", MODULE_NAME,
+					error->message);
+				g_clear_pointer(&ret, g_variant_unref);
+				return false;
+			}
+			g_clear_pointer(&ret, g_variant_unref);
+		}
+		claimed = claim;
+	} else {
+		claimed = false;
+	}
+	
+	return true;
+}
+
+static void display_state_trigger(gconstpointer data)
+{
+	(void)data;
+	display_state = datapipe_get_gint(display_state_pipe);
+	iio_als_claim_light_sensor(display_state == MCE_DISPLAY_ON);
+}
+
+static void iio_als_properties_changed(GDBusProxy * proxy,
+				       GVariant * changed_properties, GStrv invalidated_properties, gpointer user_data)
+{
+
+	GVariantDict dict;
+
+	g_variant_dict_init(&dict, changed_properties);
+
+	if (g_variant_dict_contains(&dict, "LightLevel")) {
+
+		int ilux = (int)iio_als_get_light_value(iio_proxy);
+		(void)execute_datapipe(&light_sensor_pipe, GINT_TO_POINTER(ilux), USE_INDATA, CACHE_INDATA);
+
+	}
+
+	g_variant_dict_clear(&dict);
+}
+
+static void iio_als_sensors_appeared(GDBusConnection * connection, const gchar * name, const gchar * name_owner,
+				     gpointer user_data)
+{
+	GError *error = NULL;
+	GVariant *ret = NULL;
+	(void)name;
+	(void)name_owner;
+	(void)connection;
+	(void)user_data;
+
+	mce_log(LL_INFO, "%s: Found iio_sensor_proxy", MODULE_NAME);
+
+	iio_proxy = g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SYSTEM,
+						  G_DBUS_PROXY_FLAGS_NONE,
+						  NULL,
+						  "net.hadess.SensorProxy",
+						  "/net/hadess/SensorProxy", "net.hadess.SensorProxy", NULL, NULL);
+
+	g_signal_connect(G_OBJECT(iio_proxy), "g-properties-changed", G_CALLBACK(iio_als_properties_changed), NULL);
+
+	if (display_state == MCE_DISPLAY_ON)
+		iio_als_claim_light_sensor(true);
+}
+
+static void iio_als_sensors_vanished(GDBusConnection * connection, const gchar * name, gpointer user_data)
+{
+	(void)name;
+	(void)user_data;
+	(void)connection;
+	if (iio_proxy) {
+		g_clear_object(&iio_proxy);
+		iio_proxy = NULL;
+		mce_log(LL_WARN, "%s: connection to iio_sensor_proxy lost", MODULE_NAME);
+		iio_als_claim_light_sensor(false);
+	}
+}
+
+G_MODULE_EXPORT const char *g_module_check_init(GModule * module);
+const char *g_module_check_init(GModule * module)
+{
+	(void)module;
+
+	mce_log(LL_DEBUG, "Initalizing %s", MODULE_NAME);
+
+	cal_scale = mce_conf_get_int("IioAls", "CalScale", 1000, NULL);
+	if (cal_scale < 0) 
+		cal_scale = 1000;
```

Comment:
Should this perhaps print a warning, when we don't get a scale? (Not sure just wondering)

Reply:
no the kernel provides the scale, and iio-sensor-proxy apples it. This option is to be used only if the kernel is lying to us, like on d4, so in fact we expect this option to _not_ provide a scale in most cases. the comments in mce.ini explains this